modules/vmfcore/src/statistics.cpp

```diff
@@ -0,0 +1,1276 @@
+                    break;
+                case Variant::type_real:
+                    m_value = Variant( m_value.get_real() - fieldValue.get_real() );
+                    break;
+                default:
+                    VMF_EXCEPTION( vmf::NotImplementedException, "Operation not applicable to this data type" );
+                }
+                break;
+            }
+            return true;
+        }
+    virtual Variant value() const
+        {
+            std::unique_lock< std::mutex > lock( m_lock );
+            return m_value;
+        }
+
+private:
+    mutable std::mutex m_lock;
+    Variant m_value;
+
+public:
+    static StatOpBase* createInstance()
+        { return new StatOpSum(); }
+};
+
+class StatOpLast: public StatOpBase
+{
+public:
+    StatOpLast()
+        {}
+    virtual ~StatOpLast()
+        {}
+
+public:
+    virtual std::string name() const
+        { return StatOpFactory::builtinName( StatOpFactory::BuiltinOp::Last ); }
+    virtual void reset()
+        {
+            std::unique_lock< std::mutex > lock( m_lock );
+            m_value = Variant();
+        }
+    virtual bool handle( StatAction::Type action, const Variant& fieldValue )
+        {
+            std::unique_lock< std::mutex > lock( m_lock );
+            switch( action )
+            {
+            case StatAction::Add:
+                m_value = fieldValue;
+                break;
+            case StatAction::Remove:
+                return false;
+            }
+            return true;
+        }
+    virtual Variant value() const
+        {
+            std::unique_lock< std::mutex > lock( m_lock );
+            return m_value;
+        }
+
+private:
+    mutable std::mutex m_lock;
+    Variant m_value;
+
+public:
+    static StatOpBase* createInstance()
+        { return new StatOpLast(); }
+};
+
+// class StatOpFactory
+
+StatOpBase* StatOpFactory::create( const std::string& name )
+{
+    // TODO: use lock for thread-safe access to UserOpMap instance
+    const UserOpMap& ops = getClassMap();
+
+    auto it = ops.find( name );
+    if( it == ops.end() )
+    {
+        VMF_EXCEPTION( vmf::NotFoundException, "User operation not registered: '" + name + "'" );
+    }
+
+    StatOpBase* op = (it->second)();
+    if( op == nullptr )
+    {
+        VMF_EXCEPTION( vmf::NullPointerException, "User operation isn't created: '" + name + "'" );
+    }
+
+    return op;
+}
+
+void StatOpFactory::registerUserOp( InstanceCreator createInstance )
+{
+    if( createInstance == nullptr )
+    {
+        VMF_EXCEPTION( vmf::NullPointerException, "Null pointer to the class instance creator for a user operation" );
+    }
+
+    std::unique_ptr< StatOpBase > userOp( createInstance() );
+    std::string userOpName = userOp->name();
+
+    // TODO: use lock for thread-safe access to UserOpMap instance
+    UserOpMap& ops = getClassMap();
+
+    auto it = ops.find( userOpName );
+    if( it != ops.end() )
+    {
+        if( it->second != createInstance )
```

Comment:
I would treat as a error a duplicate registration even with the same `createInstance` since this looks like a bug in application code logic